File: sample-videochat-webrtc/sample-videochat-webrtc/Components/Call/CallHelper/CallHelper.m
//
//  CallHelper.m
//  sample-videochat-webrtc
//
//  Created by Injoit on 20.07.2020.
//  Copyright © 2020 QuickBlox Team. All rights reserved.
//

#import "CallHelper.h"
#import "SessionsController.h"
#import "Log.h"
#import "CallPayload.h"
#import "MediaListener.h"
#import "MediaController.h"
#import "NSDate+Videochat.h"

@interface CallHelper ()

@property (nonatomic, strong) CallKitManager *callKit;
@property (nonatomic, strong) SessionsController *sessionsController;

@end

@interface CallHelper (ActiveSession) <SessionsControllerDelegate>
@end

@interface CallHelper (CallKit) <CallKitManagerDelegate>
@end


@implementation CallHelper

- (instancetype)init {
    self = [super init];
    if (self) {
        self.callKit = [CallKitManager new];
        self.callKit.delegate = self;
        self.sessionsController = [SessionsController new];
        self.sessionsController.delegate = self;
    }
    return self;
}

- (MediaController *)generateMediaController {
    MediaController *mediaController = [[MediaController alloc] init];
    mediaController.delegate = self.sessionsController;
    self.callKit.actionDelegate = mediaController;
    return mediaController;
}

- (MediaListener *)generateMediaListener {
    MediaListener *mediaListener = [[MediaListener alloc] init];
    self.sessionsController.mediaListenerDelegate = mediaListener;
    return mediaListener;
}

- (NSString *)registeredCallId {
    return self.sessionsController.activeSessionId;
}

- (BOOL)callReceivedWithSessionId:(NSString *)sessionID {
    return [self.sessionsController session:sessionID confirmToState:SessionStateReceived];
}

- (void)registerCallWithPayload:(NSDictionary *)payload
                     completion:(void (^)(void))completion {
    CallPayload *call = [[CallPayload alloc] initWithPayload:payload];
    IncommingCallState state = IncommingCallStateValid;
    if (call.valid == NO) {
        state = IncommingCallStateInvalid;
    } else if (call.missed || [self.sessionsController session:call.sessionID confirmToState:SessionStateRejected]) {
        state = IncommingCallStateMissed;
    }
    if (self.callKit.callUUID && [self.sessionsController session:call.sessionID confirmToState:SessionStateNew])  {
        // when self.callKit.callUUID != nil
        // at that moment has the active call
        Log(@"[%@] Received a voip push with another session that has an active call at that moment.", NSStringFromClass(CallHelper.class));
        [self.sessionsController reject:call.sessionID userInfo:nil];
        return;
    }
    
    [self.callKit reportIncomingCall:call.sessionID
                               title:call.title
                            hasVideo:call.hasVideo
                               state:state
                          completion:completion];
    
    if (state != IncommingCallStateValid) {
        return;
    }
    
    [self.sessionsController activate:call.sessionID timestamp:call.timestamp];
    
    if ([self.delegate respondsToSelector:@selector(helper:didRegisterCall:mediaListener:mediaController:direction:members:hasVideo:)]) {
        [self.delegate helper:self
              didRegisterCall:call.sessionID
                mediaListener:[self generateMediaListener]
               mediaController:[self generateMediaController]
                    direction:CallDirectionIncoming
                      members:call.members
                     hasVideo:call.hasVideo];
    }
}

- (void)registerCallWithMembers:(NSDictionary<NSNumber *, NSString *>*)members
                       hasVideo:(BOOL)hasVideo {
    // Prepare call
    NSDictionary *payload = [self.sessionsController activateWithMembers:members
                                                                          hasVideo:hasVideo];
    if (!payload.count) {
        Log(@"[%@] You should login to use VideoChat API. Session hasn’t been created. Please try to relogin.", NSStringFromClass(CallHelper.class));
        return;
    }
        
    CallPayload *call = [[CallPayload alloc] initWithPayload:payload];
    
    //Showing CallKit screen
    [self.callKit reportOutgoingCall:call.sessionID
                               title:call.title
                            hasVideo:call.hasVideo
                          completion:nil];
    
    //Sending VOIP call event
    NSData *data = [NSJSONSerialization dataWithJSONObject:payload
                                                   options:NSJSONWritingPrettyPrinted
                                                     error:nil];
    NSString *message = [[NSString alloc] initWithData:data
                                              encoding:NSUTF8StringEncoding];
    
    //Determine participants who are offline to send them a VOIP Push
    for (NSNumber *member in members.allKeys) {
        [QBChat.instance pingUserWithID:member.unsignedIntValue timeout:3.0f completion:^(NSTimeInterval timeInterval, BOOL success) {
            if (success) {
                Log(@"[%@] Participant with id: %@ is online. There is no need to send a VoIP notification.",  NSStringFromClass(CallHelper.class), member);
                return;
            }
            QBMEvent *event = [QBMEvent event];
            event.notificationType = QBMNotificationTypePush;
            event.usersIDs = [NSString stringWithFormat:@"%@", member];
            event.type = QBMEventTypeOneShot;
            event.message = message;
            
            [QBRequest createEvent:event
                      successBlock:^(QBResponse *response, NSArray<QBMEvent *> *events) {
                Log(@"[%@] Send voip push to Participant with id: %@ - Success",  NSStringFromClass(CallHelper.class), member);
            } errorBlock:^(QBResponse * _Nonnull response) {
                
                Log(@"[%@] Send voip push to Participant with id: %@  - Error %@",  NSStringFromClass(CallHelper.class), member, response.error.description);
            }];
        }];
    }
    
    // Start call
    NSString *timeStamp = [NSString stringWithFormat:@"%f", [[NSDate date] currentTimestamp]];
    NSDictionary *userInfo = @{@"timestamp" : timeStamp};
    [self.sessionsController start:call.sessionID userInfo:userInfo];
    if ([self.delegate respondsToSelector:@selector(helper:didRegisterCall:mediaListener:mediaController:direction:members:hasVideo:)]) {
        [self.delegate helper:self
                  didRegisterCall:call.sessionID
                mediaListener:[self generateMediaListener]
               mediaController:[self generateMediaController]
                        direction:CallDirectionOutgoing
                          members:call.members
                         hasVideo:call.hasVideo];
    }
}

- (void)unregisterCall:(NSString *)callId
              userInfo:(NSDictionary<NSString *, NSString *>*)userInfo {
    [self.sessionsController reject:callId
                               userInfo:userInfo];
}

- (void)updateCall:(NSString *)callId title:(NSString *)title {
    [self.callKit reportUpdateCall:callId title:title];
}

@end

@implementation CallHelper (ActiveSession)

- (void)controller:(nonnull SessionsController *)controller didAcceptSession:(nonnull NSString *)sessionId {
    [self.callKit reportAcceptCall:sessionId];
}

- (void)controller:(nonnull SessionsController *)controller didEndWaitSession:(nonnull NSString *)sessionId {
    if ([self.delegate respondsToSelector:@selector(helper:didUnregisterCall:)]) {
        [self.delegate helper:self didUnregisterCall:sessionId];
    }
    [self.callKit reportEndCall:sessionId reason:CXCallEndedReasonUnanswered];
}

- (void)controller:(SessionsController *)controller
didCloseSession:(NSString *)sessionId {
    if ([self.delegate respondsToSelector:@selector(helper:didUnregisterCall:)]) {
        [self.delegate helper:self didUnregisterCall:sessionId];
    }
    [self.callKit reportEndCall:sessionId];
}

- (void)controller:(nonnull SessionsController *)controller didChangeAudioState:(BOOL)enabled session:(nonnull NSString *)sessionId {
    [self.callKit muteAudio:!enabled call:sessionId];
}

- (void)controller:(nonnull SessionsController *)controller didReceiveIncomingSession:(nonnull NSDictionary *)payload {
    [self registerCallWithPayload:payload completion:^{}];
}


@end

@implementation CallHelper (CallKit)

- (void)callKit:(nonnull CallKitManager *)callKit didEndCall:(nonnull NSString *)sessionId {
    // external ending using "reportEndCall" methods
}

- (void)callKit:(nonnull CallKitManager *)callKit didTapAnswer:(nonnull NSString *)sessionId {
    [self.sessionsController accept:sessionId userInfo:nil];
    if ([self.delegate respondsToSelector:@selector(helper:didAcceptCall:)]) {
        [self.delegate helper:self didAcceptCall:sessionId];
    }
}

- (void)callKit:(nonnull CallKitManager *)callKit didTapRedject:(nonnull NSString *)sessionId {    
    [self.sessionsController reject:sessionId userInfo:nil];
}


@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/Calls/CallHelper/CallHelper.swift
//
//  CallHelper.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 01.04.2021.
//  Copyright © 2021 QuickBlox. All rights reserved.
//

import UIKit
import Quickblox
import QuickbloxWebRTC

enum CallDirection {
    case incoming
    case outgoing
}

protocol CallHelperDelegate: AnyObject {
    func helper(_ helper: CallHelper, didAcceptCall callId: String)
    func helper(_ helper: CallHelper, didUnregisterCall callId: String)
    func helper(_ helper: CallHelper,
                didRegisterCall callId: String,
                mediaListener: MediaListener,
                mediaController: MediaController,
                direction: CallDirection,
                members: [NSNumber: String],
                hasVideo: Bool)
}

class CallHelper: NSObject {
    //MARK: - Properties
    weak var delegate: CallHelperDelegate?
    var registeredCallId: String? {
        return sessionsController.activeSessionId
    }
    private let sessionsController = SessionsController()
    private let callKit = CallKitManager()
    
    // MARK: - Life Cycle
    override init() {
        super.init()
        
        callKit.delegate = self
        sessionsController.delegate = self
    }
    
    //MARK: - Public Methods
    func callReceived(_ sessionID: String) -> Bool {
        return sessionsController.session(sessionID, confirmToState: .received)
    }
    
    func registerCall(withPayload payload: [String : String], completion: (() -> Void)? = nil) {
        let call = CallPayload(payload: payload)
        var state: IncommingCallState = .valid
        if call.valid == false {
            state = .invalid
        } else if call.missed || sessionsController.session(call.sessionID, confirmToState: .rejected) {
            state = .missed
        }
        if let callUUID = callKit.callUUID(), callUUID.uuidString != call.sessionID, sessionsController.session(call.sessionID, confirmToState: .new) {
            // when self.callKit.callUUID != nil
            // at that moment has the active call
            debugPrint("\(#function) Received a voip push with another session that has an active call at that moment")
            sessionsController.reject(call.sessionID, userInfo: nil)
            return
        }
        callKit.reportIncomingCall(sessionId: call.sessionID, title: call.title, hasVideo: call.hasVideo, state: state, completion: completion)
        if state != .valid {
            return
        }
        sessionsController.activate(call.sessionID, timestamp: Int64(call.timestamp))
        delegate?.helper(self, didRegisterCall: call.sessionID,
                         mediaListener: generateMediaListener(),
                         mediaController: generateMediaController(),
                         direction: .incoming,
                         members: call.members,
                         hasVideo: call.hasVideo)
    }
    
    
    func registerCall(withMembers members: [NSNumber: String], hasVideo:Bool, userInfo: [String: String]?) {
        // Prepare call
        let payload = sessionsController.activateNewSession(withMembers: members, hasVideo: hasVideo)
        
        if payload.isEmpty {
            debugPrint("\(#function) You should login to use VideoChat API. Session hasn’t been created. Please try to relogin.")
            return
        }
        
        let call = CallPayload(payload: payload)
        
        //Showing CallKit screen
        callKit.reportOutgoingCall(sessionId: call.sessionID, title: call.title, hasVideo: call.hasVideo, completion: nil)
        
        //Sending VOIP call event
        let data = try? JSONSerialization.data(withJSONObject: payload,
                                               options: .prettyPrinted)
        var message = ""
        if let data = data {
            message = String(data: data, encoding: .utf8) ?? ""
        }
        
        let arrayUserIDs = members.keys.map({"\($0)"})
        let usersIDsString = arrayUserIDs.joined(separator: ",")
        
        let event = QBMEvent()
        event.notificationType = QBMNotificationType.push
        event.usersIDs = usersIDsString
        event.type = .oneShot
        event.message = message
        QBRequest.createEvent(event, successBlock: { response, events in
            debugPrint("\(#function) Send voip push - Success")
        }, errorBlock: { response in
            debugPrint("\(#function) Send voip push Error: \(response.error?.error?.localizedDescription ?? "")")
        })
        
        // Start call
        sessionsController.start(call.sessionID, userInfo: userInfo)
        
        delegate?.helper(self, didRegisterCall: call.sessionID,
                         mediaListener: generateMediaListener(),
                         mediaController: generateMediaController(),
                         direction: .outgoing,
                         members: call.members,
                         hasVideo: call.hasVideo)
    }
    
    func unregisterCall(_ callId: String, userInfo: [String: String]?) {
        sessionsController.reject(callId, userInfo: userInfo)
    }
    
    func updateCall(_ callId: String, title: String) {
        callKit.reportUpdateCall(sessionId: callId, title: title)
    }
    
    //MARK: - Private Methods
    private func generateMediaController() -> MediaController {
        let mediaController = MediaController()
        mediaController.delegate = sessionsController
        callKit.actionDelegate = mediaController
        return mediaController
    }

    private func generateMediaListener() -> MediaListener {
        let mediaListener = MediaListener()
        sessionsController.mediaListenerDelegate = mediaListener
        return mediaListener
    }
}

//MARK: - CallKitManagerDelegate
extension CallHelper: CallKitManagerDelegate {
    func callKit(_ callKit: CallKitManager, didTapAnswer sessionId: String) {
        sessionsController.accept(sessionId, userInfo: nil)
        delegate?.helper(self, didAcceptCall: sessionId)
    }
    
    func callKit(_ callKit: CallKitManager, didTapRedject sessionId: String) {
        sessionsController.reject(sessionId, userInfo: ["rejectCall": "CallKit"])
    }
    
    func callKit(_ callKit: CallKitManager, didEndCall sessionId: String) {
        // external ending using "reportEndCall" methods
    }
}

//MARK: - CallSessionDelegate
extension CallHelper: CallSessionDelegate {
    func controller(_ controller: SessionsController, didChangeAudioState enabled: Bool, forSession sessionId: String) {
        callKit.muteAudio(!enabled, forCall: sessionId)
    }
    
    func controller(_ controller: SessionsController, didReceiveIncomingSession payload: [String : String]) {
        registerCall(withPayload: payload, completion: nil)
    }
    
    func controller(_ controller: SessionsController, didCloseSession sessionId: String, userInfo: [String : String]?) {
        delegate?.helper(self, didUnregisterCall: sessionId)
        callKit.reportEndCall(sessionId: sessionId)
    }
    
    func controller(_ controller: SessionsController, didEndWaitSession sessionId: String) {
        delegate?.helper(self, didUnregisterCall: sessionId)
        callKit.reportEndCall(sessionId: sessionId, reason: .unanswered)
    }
    
    func controller(_ controller: SessionsController, didAcceptSession sessionId: String) {
        callKit.reportAcceptCall(sessionId: sessionId)
    }
}
